### PR TITLE
Support Abstract Properties

### LINKED ISSUE #125: [Task] Support `bamm:refines` replacement
**Is your task related to a problem? Please describe.**
In OpenManufacturingPlatform/sds-bamm-aspect-meta-model#34, the `bamm:extends` keyword and the concept of Abstract Entities was introduced to replace the complicated `bamm:refines` construct. Since the concept of Abstract Entities alone is not sufficient to replace `bamm:refines`, OpenManufacturingPlatform/sds-bamm-aspect-meta-model#110 proposes alternative additional modeling techniques. Whatever is used as a replacement needs to be supported in the SDK.

**Describe the solution you'd like**
Implement loading/handling the corresponding model changes

### PR DESCRIPTION
**Pull request by atextor**
Fixes #125. This implements the `bamm:AbstractProperty` and proper corresponding
semantics of `bamm:extends`.

### CODE REVIEW

File: core/sds-aspect-model-document-generators/src/main/java/io/openmanufacturing/sds/aspectmodel/generator/jsonschema/AspectModelJsonSchemaVisitor.java [hunk 1]
```diff
@@ -239,18 +242,38 @@ public JsonNode visitHasProperties( final HasProperties element, final ObjectNod
       return context;
    }
 
+   private Characteristic determineCharacteristic( final Property property ) {
+      if ( property.getCharacteristic() != null ) {
```
Comment (georgschmidtdumont) [inline core/sds-aspect-model-document-generators/src/main/java/io/openmanufacturing/sds/aspectmodel/generator/jsonschema/AspectModelJsonSchemaVisitor.java:246]: In which scenario is the Characteristic `null`?

File: core/sds-aspect-model-java-generator/src/test-shared/java/io/openmanufacturing/sds/test/shared/compiler/JavaCompiler.java [hunk 2]
```diff
@@ -45,19 +56,45 @@ public static Map<QualifiedName, Class<?>> compile( final List<QualifiedName> lo
             .map( key -> new CompilerInput( key.toString(), sources.get( key ) ) )
             .collect( Collectors.toList() );
 
+      if ( System.getProperty( WRITE_SOURCES_PROPERTY ) != null ) {
```
Comment (georgschmidtdumont) [inline core/sds-aspect-model-java-generator/src/test-shared/java/io/openmanufacturing/sds/test/shared/compiler/JavaCompiler.java:59]: Why did you make this optional? To me it sound really helpful to have the generated classes from the tests written to an explicit directory.

File: core/sds-aspect-model-java-generator/src/test-shared/java/io/openmanufacturing/sds/test/shared/compiler/JavaCompiler.java [hunk 3]
```diff
@@ -45,19 +56,45 @@ public static Map<QualifiedName, Class<?>> compile( final List<QualifiedName> lo
...
       };
 
       final List<String> compilerOptions = List.of( "-classpath", System.getProperty( "java.class.path" ) );
-      compiler.getTask( null, manager, diagnosticListener, compilerOptions, null, compilerInput ).call();
+      compiler.getTask( null, manager, diagnosticListener, compilerOptions, null, compilerInput ).
+
+            call();
 
       return loadOrder.stream()
-            .collect( Collectors.toMap( Function.identity(), qualifiedName -> defineAndLoad( qualifiedName, manager ) ) );
```
Comment (georgschmidtdumont) [inline core/sds-aspect-model-java-generator/src/test-shared/java/io/openmanufacturing/sds/test/shared/compiler/JavaCompiler.java:60]: Please reformat.

**Review by georgschmidtdumont — CHANGES_REQUESTED** [review]

File: core/sds-aspect-meta-model-java/src/main/java/io/openmanufacturing/sds/metamodel/loader/MetaModelBaseAttributes.java [hunk 4]
```diff
@@ -145,56 +143,80 @@ public static MetaModelBaseAttributes from( final KnownVersion metaModelVersion,
...
-            .substring( 0, 7 );
-      return ImmutableList.copyOf( model.listStatements( modelElement, RDF.type, (RDFNode) null ) ).stream()
-            .findFirst()
-            .filter( statement -> statement.getObject().isResource() )
-            .map( statement -> statement.getObject().asResource() )
-            .map( type -> metaModelResourceResolver.getAspectModelUrn( type.getURI() ) )
-            .flatMap( Value::toJavaOptional )
-            .map( AspectModelUrn::getName )
-            .orElse( "" ) + randomPart;
+   private static Optional<String> getName( final Resource modelElement, final BAMM bamm ) {
```
[the remark this answers is not in the document]
Comment (atextor) [reply core/sds-aspect-meta-model-java/src/main/java/io/openmanufacturing/sds/metamodel/loader/MetaModelBaseAttributes.java:184]: They were in fact the same method, one of them was not used. I suppose this was either an error during refactoring or merging. The unused (getNameFromInheritanceTree) is removed now.

File: core/sds-aspect-model-document-generators/src/main/java/io/openmanufacturing/sds/aspectmodel/generator/jsonschema/AspectModelJsonSchemaVisitor.java (hunk 1 above)
Comment (atextor) [reply core/sds-aspect-model-document-generators/src/main/java/io/openmanufacturing/sds/aspectmodel/generator/jsonschema/AspectModelJsonSchemaVisitor.java:246]: The Characteristic can be `null` if the Property is abstract.

File: core/sds-aspect-model-java-generator/src/test-shared/java/io/openmanufacturing/sds/test/shared/compiler/JavaCompiler.java (hunk 2 above)
Comment (atextor) [reply core/sds-aspect-model-java-generator/src/test-shared/java/io/openmanufacturing/sds/test/shared/compiler/JavaCompiler.java:59]: Yes, I would like to have this as well. The problem is that many model test models define model elements with the same name (e.g. "TestEntity") but with different structure and semantics. When you selectively run one test, this is no problem. When you run the test suite, where code for all test models is generated, you end up in a situation where one test overwrites previously generated code from another test.
Let's say AspectA has a TestEntity and AspectB also has TestEntity but with a different structure. Then AspectB's TestEntity's constructor does not match the `new TestEntity` cann in AspectA and you end up with compilation errors. We would have to cleanly move each test model into its own namespace. I consider this out of scope for this PR though.

File: core/sds-aspect-model-document-generators/src/main/java/io/openmanufacturing/sds/aspectmodel/generator/jsonschema/AspectModelJsonSchemaVisitor.java (hunk 1 above)
Comment (georgschmidtdumont) [reply core/sds-aspect-model-document-generators/src/main/java/io/openmanufacturing/sds/aspectmodel/generator/jsonschema/AspectModelJsonSchemaVisitor.java:246]: Would it not make sense to define the Characteristic as optional in the `Property` class?

Comment (atextor) [reply core/sds-aspect-model-document-generators/src/main/java/io/openmanufacturing/sds/aspectmodel/generator/jsonschema/AspectModelJsonSchemaVisitor.java]: You're correct. I have now updated `getCharacteristic()` in Property to return `Optional<Characteristic>`. This communicates in a clearer way that there are cases in which no Characteristic is present.

**Review by georgschmidtdumont — APPROVED** [review]

**Pull request merged**